enhancements/builds/node-cache-for-builds.md

```diff
@@ -0,0 +1,164 @@
+---
+title: node-cache-for-builds
+authors:
+  - "@adambkaplan"
+reviewers:
+  - "@nalind
+  - "@rhdan"
+  - "@TomSweeneyRedHat"
+  - "@bparees"
+approvers:
+  - "@bparees"
+creation-date: 2019-02-18
+last-updated: 2019-02-18
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Node Cache for Builds
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in
+  [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+None.
+
+## Summary
+
+Builds in OpenShift 4.x currently require every layer of the base image to be pulled from its source
+registry every time the build is run. In OCP 3.x builds could take advantage of image layers that
+were stored on the build's node. This proposal provides a means of partially restoring this
+node-level cache by allowing build pods to avoid pulling layers that are present on the node. Unlike
+3.x, however, layers created during the build process will not be directly cached on the node due to
+buildah's isolation.
+
+## Motivation
+
+### Goals
+
+1. Provide a means for builds to take advantage of image layers that are stored on the node.
+
+### Non-Goals
+
+1. Provide a means for builds to write their layers to a cache (node or shared).
+2. Provide a means for builds to share an image layer cache within a namespace.
+3. Provide a means for the internal registry to pre-populate an image layer cache within a namespace
+   or across a cluster.
+4. Add persistent volume support for builds.
+
+## Proposal
+
+Build pods will mount the image layer cache that exists on every OpenShift node into the build pod.
+Buildah - running within the build pod - will use this layer cache to reduce the number of layers
+that need to be pulled when the build commences.
+
+### User Story
+
+As a developer using OpenShift to build container images I want builds to use image layers cached on
+the node So that builds don't have to pull all layers in their base images And builds run faster.
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. The build controller will mount the node image cache located at `/var/lib/containers/storage`
+   into the build pod as `readOnly`. The mount point should not conflict with buildah's own default
+   storage directory. Ex: `/var/lib/shared/containers/storage`.
+2. The openshift/builder image is modified to configure additional image stores for buildah (via
+   `/etc/containers/storage.conf`).
```

Comment:
To echo the planning discussion again around enabling as much as we can tekton/buildv2 with all this, if the processing of `/etc/containers/storage.conf` can be done in a new utility repo that can be imported by openshift/builder and any tekton utilities / build v2 operators, etc. (vs. those operators / utilities having to import openshift/builder for any common code), that would be ideal.

When constructing the common code, some sort of abstraction layer to deal with seeding buildah CLI vs. API in the different targets is warranted.

Reply:
FWIW I've called out a "common utility repo" in the entitlements proposal as well.

Reply:
@gabemontero So today we bake in a `storage.conf` file into the openshift/builder image, and right now we don't modify this file via the build controller [1]. The additional stores are in fact already configured - we may not need to do anything to this file as a part of the enhancement [2].

For Tekton/builds v2, the images which contain buildah will need to likewise be updated to have a default `storage.conf` that could take advantage of the additional store (with the appropriate mounts and buildah CLI options).

[1] https://github.com/openshift/openshift-controller-manager/blob/master/pkg/build/controller/strategy/util.go#L382-L411
[2] https://github.com/openshift/builder/blob/master/imagecontent/storage.conf#L19-L23

Reply:
Makes sense @adambkaplan ... and I see you've pushed a commit that calls out the need for both buildah CLI and API updates to leverage this.  LGTM.

I'll ping the app svcs/dev tools tekton guys about this in slack.  Certainly feel free to tag them in this review as you see fit.

Reply:
makes sense @bparees 

I think details / lessons learned from this work, plus ensuring any buildah API enhancements around specifying storage are available from the CLI (and on the suface I don't see why they would do it just for API vs. CLI), would be a sufficient starting point for tekton based solutions built by our brethren on that side of the shop.